### PR TITLE
[chainweb-node Main] add systemd watchdog support.

### PR DESCRIPTION
**Pull request by gregorycollins**

### CODE REVIEW

File: src/Chainweb/Utils.hs [hunk 1]
```diff
@@ -877,16 +878,19 @@ runForeverThrottled logfun name burst rate a = mask $ \umask -> do
...
-leadingZeros b = int (B.length x) * 8 + case B.uncons y of
-    Just (h, _) -> int $ countLeadingZeros h
-    Nothing -> 0
-  where
-    (x, y) = B.span (== 0x00) b
+-- Count leading zero bits of a bytestring
+
+-- | Count leading zero bits of a bytestring
+--
+leadingZeros :: Integral int => B.ByteString -> int
```
Comment (gregorycollins) [inline src/Chainweb/Utils.hs]: I should probably move this hunk into its own PR -- I noticed it while I was reading this. Will delete

File: chainweb.cabal [hunk 2]
```diff
@@ -367,6 +372,9 @@ library
 
     if flag(ed25519)
         cpp-options: -DWITH_ED25519=1
+    if flag(use_systemd)
```
Comment (gregorycollins) [inline chainweb.cabal]: I forgot to add an `&& os(linux)` here

File: stack.yaml [hunk 3]
```diff
@@ -14,6 +14,7 @@ extra-deps:
   - paths-0.2.0.0
   - streaming-events-1.0.0
   - strict-tuple-0.1.2
+  - systemd-1.2.0
```
Comment (fosskers) [inline stack.yaml]: @emilypi ran a `stack build` on her Mac with this - and everything compiled fine.

File: default.nix [hunk 4]
```diff
@@ -50,6 +51,17 @@ in
...
 
+        systemd =
+          if systemdFeatures then
+            callHackageDirect {
+              pkg = "systemd";
+              # we'd like to use a newer one but it requires network >= 3
+              ver = "1.2.0";
+              sha256 = "1mwrrki3zsc4ncr7psjv9iqkzh7f25c2ch4lf2784fh6q46i997j";
+            }
+          else null;
```
Comment (fosskers) [inline default.nix]: `null` is okay?

File: src/Chainweb/Utils/Watchdog.hs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        either return impossible
+  where
+    impossible _ = fail "impossible: watchdog returned first"
+
+    watchdogTimer :: Int -> IO ()
+    watchdogTimer usec0 = forever $ do
+        void $ tryAllSynchronous $ SD.notifyWatchdog
+        threadDelay usec
+      where
+        usec = if usec0 <= 0 then defaultWatchdogUsec else usec0
```
Comment (fosskers) [inline src/Chainweb/Utils/Watchdog.hs]: Any potential issues here the `usec` is *just* too close to what the external `systemd` is expecting?

File: default.nix (hunk 4 above)
Comment (gregorycollins) [reply default.nix]: The osx nix build wouldn't have passed otherwise

File: src/Chainweb/Utils/Watchdog.hs (hunk 5 above)
Comment (gregorycollins) [reply src/Chainweb/Utils/Watchdog.hs]: In general no -- typical values for watchdog in seconds are 30s - 2m
